File: backend/models/lesson.py
from datetime import datetime
from typing import Optional

from sqlalchemy import BigInteger, DateTime, ForeignKey, Integer, String, func
from sqlalchemy.orm import Mapped, mapped_column

from database import Base


class Lesson(Base):
    __tablename__ = "lessons"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    course_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("courses.id"), nullable=False
    )
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    week_number: Mapped[int] = mapped_column(Integer, nullable=False)
    material_id: Mapped[Optional[int]] = mapped_column(
        BigInteger, ForeignKey("materials.id"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

File: frontend/src/lib/api/types/lesson.ts
/** Lesson row from `GET/POST/PATCH /courses/.../lessons` and `GET /lessons/{id}`. */
export type Lesson = {
  id: number
  course_id: number
  title: string
  week_number: number
  material_id: number | null
  created_at: string
}

export type CreateLessonBody = {
  title: string
  week_number: number
}

export type UpdateLessonBody = {
  title?: string
  week_number?: number
}
